File: RootTools2/src/main/java/com/rarnu/tools/neo/xposed/MIUITheme.java
package com.rarnu.tools.neo.xposed;

import android.content.Context;
import android.content.Intent;
import com.rarnu.tools.neo.utils.FileUtils;
import de.robv.android.xposed.*;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import java.io.File;

public class MIUITheme implements IXposedHookZygoteInit, IXposedHookLoadPackage {

    public static Object getDrmResultSUCCESS() {
        try {
            Class<Enum> clsEnum = (Class<Enum>) Class.forName("miui.drm.DrmManager$DrmResult");
            if (clsEnum != null) {
                return Enum.valueOf(clsEnum, "DRM_SUCCESS");
            }
        } catch (Throwable th) {
            XposedBridge.log(th.toString());

        }
        return null;
    }

    @Override
    public void initZygote(StartupParam startupParam) throws Throwable {
        XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        prefs.reload();
        if (!prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
            return;
        }
        MIUI_DRM();
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        prefs.reload();
        if (lpparam.packageName.equals("miui.drm") || lpparam.packageName.equals("com.miui.system") || lpparam.packageName.equals("miui.system")) {
            if (prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
                MIUI_DRM();
            }
            return;
        }
        if (lpparam.packageName.equals("com.android.thememanager")) {
            if (prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
                XpUtils.findAndHookMethod("com.android.thememanager.util.ThemeOperationHandler", lpparam.classLoader, "isTrialable", XC_MethodReplacement.returnConstant(false));
                XpUtils.findAndHookMethod("com.android.thememanager.util.ThemeOperationHandler", lpparam.classLoader, "isLegal", XC_MethodReplacement.returnConstant(true));
                XpUtils.findAndHookMethod("com.android.thememanager.util.ThemeOperationHandler", lpparam.classLoader, "isAuthorizedResource", XC_MethodReplacement.returnConstant(true));
                XpUtils.findAndHookMethod("com.android.thememanager.util.ThemeOperationHandler", lpparam.classLoader, "isPermanentRights", XC_MethodReplacement.returnConstant(true));
            }
            if (prefs.getBoolean(XpStatus.KEY_KEEP_MTZ, false)) {
                XpUtils.findAndHookMethod("com.android.thememanager.model.ResourceOnlineProperties", lpparam.classLoader, "setDownloadPath", String.class, new XC_MethodHook() {
                    @Override
                    protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                        String oriPath = (String) param.args[0];
                        String newPath = oriPath + "1.mtz";
                        XposedBridge.log("setDownloadPath => " + newPath);
                        try {
                            FileUtils.copyFile(oriPath, newPath);
                            XposedBridge.log("setDownloadPath copy file => succ");
                            param.args[0] = newPath;
                        } catch (Exception e) {
                            XposedBridge.log("setDownloadPath error => " + e.toString());
                        }

                    }
                });
            }
        }
    }

    private void MIUI_DRM() {
        XpUtils.findAndHookMethod("miui.drm.DrmManager", "isLegal", new Object[]{Context.class, File.class, File.class, XC_MethodReplacement.returnConstant(getDrmResultSUCCESS())});
        XpUtils.findAndHookMethod("miui.drm.DrmManager", "isLegal", new Object[]{Context.class, String.class, File.class, XC_MethodReplacement.returnConstant(getDrmResultSUCCESS())});
        XpUtils.findAndHookMethod("miui.drm.DrmManager", "isLegal", new Object[]{Context.class, String.class, "miui.drm.DrmManager$RightObject", XC_MethodReplacement.returnConstant(getDrmResultSUCCESS())});
        XpUtils.findAndHookMethod("miui.drm.DrmManager", "isPermanentRights", new Object[]{File.class, XC_MethodReplacement.returnConstant(true)});
        XpUtils.findAndHookMethod("miui.drm.DrmManager", "isLegal", new Object[]{"miui.drm.DrmManager$RightObject", XC_MethodReplacement.returnConstant(true)});
        XpUtils.findAndHookMethod("miui.drm.ThemeReceiver", "validateTheme", new Object[]{XC_MethodReplacement.returnConstant(true)});
        XpUtils.findAndHookMethod("miui.content.res.ThemeRuntimeManager.ThemeReceiver", "validateTheme", new Object[]{XC_MethodReplacement.returnConstant(true)});
        XpUtils.findAndHookMethod("miui.content.res.ThemeRuntimeManager", "restoreDefault", new Object[]{new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                return null;
            }
        }});
        XpUtils.findAndHookMethod("miui.drm.ThemeReceiver", "restoreDefault", new Object[]{new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                return null;
            }
        }});
        XpUtils.findAndHookMethod("miui.content.res.ThemeRuntimeManager", "restoreDefault", new Object[]{new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                return null;
            }
        }});
        XpUtils.findAndHookMethod("miui.drm.ThemeReceiver", "onReceive", new Object[]{Context.class, Intent.class, new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                return null;
            }
        }});
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/xposed/MIUITheme.kt
@file:Suppress("Duplicates")

package com.rarnu.tools.neo.xposed

import android.content.Context
import android.content.Intent
import com.rarnu.xfunc.*
import de.robv.android.xposed.*
import java.io.File
import java.io.InputStream
import java.lang.Exception

class MIUIThemeZygote: XposedZygote() {
    override fun hook(zygote: XposedStartup) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        prefs.reload()
        if (prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
            patchDRM(null)
        }
    }
}

class MIUIThemePackage: XposedPackage() {
    override fun hook(pkg: XposedPkg) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        prefs.reload()
        if (pkg.packageName == "miui.drm" || pkg.packageName == "com.miui.system" || pkg.packageName == "miui.system") {
            if (prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
                patchDRM(pkg.classLoader)
            }
            return
        }
        if (pkg.packageName == "com.android.thememanager") {
            if (prefs.getBoolean(XpStatus.KEY_THEMECRACK, false)) {
                pkg.findClass("miui.resourcebrowser.view.ResourceOperationHandler").findMethod("isAuthorizedResource").hook { replace { result = true } }
                // payment
                pkg.findClass("miui.resourcebrowser.controller.online.NetworkHelper").findMethod("validateResponseResult", Integer.TYPE, InputStream::class.java).hook {
                    before {
                        result = args[1] as InputStream?
                    }
                }
                pkg.findClass("miui.resourcebrowser.view.ResourceOperationHandler").findMethod("onCheckResourceRightEventBeforeRealApply").hook {
                    before {
                        try {
                            thisObject.javaClass.findField("mIsLegal").setStatic(true)
                        } catch (t: Throwable) {
                        }
                    }
                }
                pkg.findClass("miui.resourcebrowser.view.ResourceOperationHandler").findMethod("onCheckResourceRightEventBeforeRealApply").hook {
                    before {
                        try {
                            thisObject.javaClass.findField("mIsLegal").setStatic(true)
                        } catch (t: Throwable) {
                        }
                    }
                }
                val clsResource = pkg.findClass("miui.resourcebrowser.model.Resource")
                if (clsResource != null) {
                    pkg.findClass("miui.resourcebrowser.controller.online.DrmService").findMethod("isLegal", clsResource).hook { replace { result = true } }
                }
                pkg.findClass("miui.resourcebrowser.view.ResourceOperationHandler").apply {
                    findMethod("isLegal").hook { replace { result = true } }
                    findMethod("onCheckResourceRightEventBeforeRealApply").hook {
                        before {
                            try {
                                thisObject.javaClass.findField("mIsLegal").setStatic(true)
                            } catch (t: Throwable) {
                            }
                        }
                    }
                }
                pkg.findClass("miui.resourcebrowser.model.Resource").findMethod("isProductBought").hook { replace { result = true } }
                pkg.findClass("miui.resourcebrowser.model.ResourceOnlineProperties").findMethod("setProductBought", java.lang.Boolean.TYPE).hook {
                    replace {
                        args = arrayOf(true)
                        result = XposedBridge.invokeOriginalMethod(method, thisObject, arrayOf(true))
                    }
                }
                pkg.findClass("miui.resourcebrowser.model.ResourceOnlineProperties").findMethod("isProductBought").hook { replace { result = true } }
                pkg.findClass("com.android.thememanager.util.ThemeOperationHandler").apply {
                    findMethod("isTrialable").hook { replace { result = false } }
                    findMethod("isLegal").hook { replace { result = true } }
                    findMethod("isAuthorizedResource").hook { replace { result = true } }
                    findMethod("isPermanentRights").hook { replace { result = true } }
                }
            }
        }
    }
}

private fun patchDRM(cl: ClassLoader?) {
    fun findClass(clsName: String) = XposedHelpers.findClass(clsName, cl)
    findClass("miui.drm.DrmManager").apply {
        findMethod("isLegal", Context::class.java, File::class.java, File::class.java).hook { replace { result= drmResultSUCCESS } }
        findMethod("isLegal", Context::class.java, String::class.java, File::class.java).hook { replace { result = drmResultSUCCESS } }
        findMethod("isPermanentRights", File::class.java).hook { replace { result = true } }
        findMethod("isRightsFileLegal", File::class.java).hook { replace { result = true } }
        try {
            val clsRightObject = Class.forName("miui.drm.DrmManager\$RightObject")
            findMethod("isLegal", Context::class.java, String::class.java, clsRightObject).hook { replace { result = drmResultSUCCESS } }
            findMethod("isPermanentRights", clsRightObject).hook { replace { result = true } }
            findMethod("isLegal", clsRightObject).hook { replace { result = drmResultSUCCESS } }
        } catch (e: Exception) {

        }
        findMethod("isSupportAd", Context::class.java).hook { replace { result = false } }
        findMethod("isSupportAd", File::class.java).hook { replace { result = false } }
        findMethod("setSupportAd", Context::class.java, java.lang.Boolean.TYPE).hook {
            before {
                args[1] = false
            }
        }
        findMethod("isLegal", File::class.java, File::class.java).hook { replace { result = drmResultSUCCESS } }
        findMethod("isLegal", String::class.java, File::class.java).hook { replace { result = drmResultSUCCESS } }
    }

    findClass("miui.drm.ThemeReceiver").apply {
        findMethod("validateTheme", Context::class.java, String::class.java, String::class.java).hook { replace { result = drmResultSUCCESS } }
        findMethod("restoreDefault").hook { replace { result = null } }
        findMethod("onReceive", Context::class.java, Intent::class.java).hook {
            before {
                val intent = args[1] as Intent?
                if (intent != null && intent.action != null && intent.action == "miui.intent.action.CHECK_TIME_UP") {
                    intent.action = ""
                    args[1] = intent
                }
            }
        }
    }
    findClass("miui.content.res.ThemeRuntimeManager\$ThemeReceiver").findMethod("validateTheme", Context::class.java, String::class.java, String::class.java).hook { replace { result = drmResultSUCCESS } }
    findClass("miui.content.res.ThemeRuntimeManager").findMethod("restoreDefault").hook { replace { result = null } }

}

val drmResultSUCCESS: Any?
    get() {
        try {
            val clsEnum = Class.forName("miui.drm.DrmManager\$DrmResult")
            val mEnum = clsEnum.getDeclaredMethod("valueOf", String::class.java)
            return mEnum.invoke(null, "DRM_SUCCESS")
        } catch (th: Throwable) {
            XposedBridge.log(th.toString())

        }
        return null
    }

